### ISSUE TITLE
twisted.web.http triggers a reactor import

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8246]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8246|
|Type|enhancement|
|Created|2016-03-23 11:57:45Z|
|Branch|https://github.com/twisted/twisted/tree/reactor-import-8246|

This is, as you can probably guess, bad, and we shouldn't do it.

It's only used in HTTPFactory.

<details><summary>Searchable metadata</summary>

```
trac-id__8246 8246
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__branches_reactor_import_8246 branches/reactor-import-8246
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1458734265376937 1458734265376937
changetime__1459088718061872 1459088718061872
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [47056]) Branching to reactor-import-8246.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Happy with this. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good. Thanks!

Only minor comments.

------

Maybe the ticket title should be updated to reflect that fact that this adds the option of an explicit reactor.

-------

Why is bad to import the reactor? Do we have documentation describing this pattern?

I think that this patter is found in many other places.

--------

Since this is not really a reactor, but a time provider, so we want to call this argument `reactor`? Maybe `clock` is closer to the usage of this variable.

-----

Do we need a separate test case for ExplicitHTTPFactoryReactor ? 


Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|amberite commented|
|-|-|

https://github.com/twisted/treq/issues/107 is somewhat related to the problem being solved here, although the reactor import occurs in twisted.web.client instead of twisted.web.http.

Is it worth making a separate ticket for each submodule of twisted.web that imports the reactor, or should the scope of this ticket be expanded to something like "twisted.web submodules import the reactor"?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Hi amberite, I think it's probably better to do submodules -- smaller pieces mean that if there's a hiccup implementing one of them, it doesn't block the others.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(also, I've noted and described the reactor import pattern in [#8256](https://github.com/twisted/twisted/issues/8256))

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47093]) Merge reactor-import-8246: Remove module-level reactor import in t.w.http

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8246](https://github.com/twisted/twisted/issues/8246)
